loki/transform/transform_array_indexing.py

```diff
@@ -496,3 +497,46 @@ def demote_variables(routine, variable_names, dimensions):
     routine.spec = Transformer(decl_map).visit(routine.spec)
 
     info(f'[Loki::Transform] Demoted variables in {routine.name}: {", ".join(variable_names)}')
+
+
+def flatten_arrays(routine, order='F', start_index=1):
+    """
+    Flatten arrays, converting multi-dimensional arrays to
+    one-dimensional arrays.
+
+    Parameters
+    ----------
+    routine : :any:`Subroutine`
+        The subroutine in which the variables should be promoted.
+    order : str
+        Assume Fortran (F) vs. C memory/array order.
+    start_index : int
+        Assume array indexing starts with `start_index`.
+    """
+    def new_dims(dim, shape):
+        if len(dim) > 1:
+            assert not isinstance(shape[-1], sym.RangeIndex)
+            _dim = [sym.Sum((dim[-2], sym.Product((shape[-2], dim[-1] - start_index))))]
```

Comment:
A little confused by this: You are checking that `shape[-1]` isn't a `RangeIndex`, yet `shape[-2]` is the dimension you are actually using here. 
I might be thinking too simple here, but to accomodate the mentioned situation of
```f90
INTEGER, INTENT(INOUT) :: x4(0:l4, 0:l3, 0:l2, 0:l1)
```
would only require to add something like
```suggestion
            if isinstance(shape[-2], sym.RangeIndex):
                _len = sym.Sum((shape[-2].upper, sym.Product((-1, shape[-2].lower))))
                _start_index = sym.Sum((start_index, shape[-2].lower))
            else:
                _len = shape[-2]
                _start_index = start_index
            if isinstance(shape[-1], sym.RangeIndex):
                _start_index = sym.Sum((start_index, shape[-1].lower))
            else:
                _start_index = start_index
            _dim = [sym.Sum((
                dim[-2],
                sym.Product((
                    _len,
                    sym.Sum((
                        dim[-1],
                        sym.Product((-1, _start_index))
                    ))
                ))
            ))]
```

I'm almost sure I'm making some mistake here between what shape to use for which, but this should take care of the possibly longer or shorter dimension due to the non-standard shape, and the offset relative to this.